File: src/boj/BOJ_01377_BubbleSort.py
import sys


class Element:
    def __init__(self, val, idx):
        self.val = val
        self.idx = idx

    def __str__(self):
        return "Value: " + str(self.val) + ", Index: " + str(self.idx)


N = int(sys.stdin.readline())
original = [None] * (N + 1)
original[0] = Element(0, 0)

for i in range(1, N + 1):
    original[i] = Element(0, 0)
    original[i].val = int(sys.stdin.readline())
    original[i].idx = i

copied = [None] * len(original)
copied[0] = Element(0, 0)
for i in range(1, N + 1):
    copied[i] = Element(0, 0)
    copied[i].val = original[i].val
    copied[i].idx = original[i].idx

original.sort(key=lambda x: x.val)

indices = [0] * (N + 1)
for i in range(1, N + 1):
    indices[i] = original[i].idx - copied[i].idx

max_val = max(indices)
print(max_val + 1)

File: src/boj/BOJ_01377_BubbleSort.java
package boj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class BOJ_01377_BubbleSort {

    private static class element implements Comparable<element>  {
        int val;
        int idx;

        @Override
        public String toString() {
            return "Value: " + val + ", Index: " + idx;
        }

        @Override
        public int compareTo(element o) {
            return Integer.compare(this.val, o.val);
        }

    }
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        element[] original = new element[N + 1];
        original[0] = new element();

        for(int i = 1; i <= N; i++) {
            original[i] = new element();
            original[i].val = Integer.parseInt(br.readLine());
            original[i].idx = i;
        }

        element[] copied = new element[original.length];
        copied[0] = new element();
        for (int i = 1; i <= N; i++) {
            copied[i] = new element();
            copied[i].val = original[i].val;
            copied[i].idx = original[i].idx;
         }
        Arrays.sort(original);

        int[] indices = new int[N + 1];
        for(int i = 1; i <= N; i++) {
            indices[i] = original[i].idx - copied[i].idx;
        }

        int max = Arrays.stream(indices).max().getAsInt();
        System.out.println(max + 1);
        
    }
}

// reference: https://scarlettb.tistory.com/114
